### ISSUE TITLE
USB storage issues (Pi 3 A+/B/B+)

### ISSUE THREAD

**Issue by dwhinham** [rsta2/circle#211]
Hi Rene,

As promised, here is my other USB issue (sorry!).

In my project we often load large (>200MB) files from SD card. Some users would like to move these files onto USB sticks for convenience (and potentially faster loading, especially on Pi 4 with USB 3.0), so I've been experimenting with USB file I/O.

On Pi 4, everything seems rock-solid, but on Pi 3, we have mixed results. There are often `Transaction failed` errors in the console or other errors such as `umsd: Cannot clear halt on endpoint 1`.

To try and narrow this down, I've written a simple test kernel ([source tree here](https://github.com/dwhinham/circle/tree/storage-test)) which reads a large file into memory and performs a SHA sum to verify it loaded correctly, a bit like our earlier SD card tests.

Here is a binary with large data file included: [Google Drive download](https://drive.google.com/file/d/1Uv-W_-PHdZJWre1P-79CrCJjDPhWvikt/view?usp=sharing)

- Inside the archive is a `testfile.bin` - this file needs to be moved to a FAT32 USB stick.
- The test is based on Circle 43.3.
- There are two 64-bit kernels for Pi 3, `kernel8-nopnp.img` expects a USB stick to be inserted from boot at all times, `kernel8-pnp.img` will ask you to insert/remove the stick for each test cycle a little bit like `39-umsdplugging`.
- The kernel will try to read the file, compare it with a known SHA256 hash, and then write it back to the stick as `testfile2.bin`.
- The non-PnP kernel will repeat this forever; the PnP kernel will let you unplug and replace the stick.

The USB stick model doesn't seem to matter; I've had similar results with various unbranded sticks, but I've mainly being testing a Sandisk Ultra USB 3.0 64GB (possibly also known as a Sandisk Cruzer?).

#### PnP kernel (FAIL)
```
logger: Circle 43.3 started on Raspberry Pi 3 Model A+ (AArch64)
00:00:00.66 timer: SpeedFactor is 1.51
00:00:00.69 sdhost: emmc1: sdhost-bcm2835 loaded
00:00:02.20 emmc: Found a valid version 3.0x SD card
00:00:03.35 dwhci: No device connected to root port
00:00:03.36 kernel: Compile time: Mar 10 2021 22:14:02
00:00:03.36 kernel: Plug in an USB flash drive!
00:00:09.27 usbdev0-1: Device ven781-5581 found
00:00:09.28 usbdev0-1: Interface int8-6-50 found
00:00:09.28 usbdev0-1: Using device/interface int8-6-50
00:00:09.44 umsd: Capacity is 58656 MByte
00:00:09.45 dwroot: Device configured
00:00:09.45 kernel: Reading 265MB from USB:...
00:00:11.26 dwhci: Transaction failed (status 0x80)
00:00:11.27 umsd: CSW transfer failed
00:00:11.27 umsd: TryRead failed
00:00:11.27 dwhci: Transaction failed (status 0x8)
00:00:11.28 umsd: Cannot clear halt on endpoint 1
00:00:11.28 kernel: Failed to read /testfile.bin into memory
```

#### Non-PnP kernel (FAIL)
```
logger: Circle 43.3 started on Raspberry Pi 3 Model A+ (AArch64)
00:00:00.66 timer: SpeedFactor is 1.51
00:00:00.69 sdhost: emmc1: sdhost-bcm2835 loaded
00:00:02.20 emmc: Found a valid version 3.0x SD card
00:00:03.10 usbdev0-1: Device ven781-5581 found
00:00:03.10 usbdev0-1: Interface int8-6-50 found
00:00:03.11 usbdev0-1: Using device/interface int8-6-50
00:00:03.26 umsd: Capacity is 58656 MByte
00:00:03.27 dwroot: Device configured
00:00:03.28 kernel: Compile time: Mar 10 2021 22:51:10
00:00:03.28 kernel: Reading 265MB from USB:...
00:00:04.07 dwhci: Transaction failed (status 0xA2)
00:00:04.08 umsd: Data transfer failed
00:00:04.08 umsd: TryRead failed
00:00:04.08 dwhci: Transaction failed (status 0x8)
00:00:04.09 umsd: Cannot clear halt on endpoint 1
00:00:04.09 kernel: Failed to read /testfile.bin into memory
```

#### PnP kernel (OK, but has errors in the log!)
```
logger: Circle 43.3 started on Raspberry Pi 3 Model B (AArch64)
00:00:00.66 timer: SpeedFactor is 1.51
00:00:00.69 sdhost: emmc1: sdhost-bcm2835 loaded
00:00:02.20 emmc: Found a valid version 3.0x SD card
00:00:03.06 usbdev0-1: Device ven424-9514, dev9-0-2 found
00:00:03.07 usbdev0-1: Interface int9-0-1 found
00:00:03.07 usbdev0-1: Function is not supported
00:00:03.08 usbdev0-1: Interface int9-0-2 found
00:00:03.08 usbdev0-1: Using device/interface int9-0-2
00:00:03.80 usbdev0-1: Device ven424-ec00 found
00:00:03.81 usbdev0-1: Using device/interface ven424-ec00
00:00:03.96 usbdev0-1: Device ven781-5581 found
00:00:03.97 usbdev0-1: Interface int8-6-50 found
00:00:03.97 usbdev0-1: Using device/interface int8-6-50
00:00:04.03 smsc951x: MAC address is B8:27:EB:78:10:CD
00:00:04.04 usbhub: Port 1: Device configured
00:00:04.19 umsd: Capacity is 58656 MByte
00:00:04.21 usbhub: Port 2: Device configured
00:00:04.21 dwroot: Device configured
00:00:04.22 kernel: Compile time: Mar 10 2021 22:14:02
00:00:04.22 kernel: Plug in an USB flash drive!
00:00:04.24 kernel: Reading 265MB from USB:...
00:00:04.48 dwhci: Transaction failed (status 0x202)
00:00:04.74 dwhci: Transaction failed (status 0x202)
00:00:05.01 dwhci: Transaction failed (status 0x202)
00:00:05.27 dwhci: Transaction failed (status 0x202)
00:00:06.04 dwhci: Transaction failed (status 0x202)
00:00:06.56 dwhci: Transaction failed (status 0x202)
00:00:06.82 dwhci: Transaction failed (status 0x202)
00:00:07.08 dwhci: Transaction failed (status 0x202)
00:00:07.34 dwhci: Transaction failed (status 0x202)
00:00:07.60 dwhci: Transaction failed (status 0x202)
00:00:07.86 dwhci: Transaction failed (status 0x202)
00:00:08.12 dwhci: Transaction failed (status 0x202)
00:00:08.38 dwhci: Transaction failed (status 0x202)
00:00:08.90 dwhci: Transaction failed (status 0x202)
00:00:09.41 dwhci: Transaction failed (status 0x202)
00:00:09.67 dwhci: Transaction failed (status 0x202)
00:00:10.19 dwhci: Transaction failed (status 0x202)
00:00:10.71 dwhci: Transaction failed (status 0x202)
00:00:11.23 dwhci: Transaction failed (status 0x202)
00:00:12.00 dwhci: Transaction failed (status 0x202)
00:00:12.26 dwhci: Transaction failed (status 0x202)
00:00:12.52 dwhci: Transaction failed (status 0x202)
00:00:12.78 dwhci: Transaction failed (status 0x202)
00:00:13.03 kernel: File read in 8.78 seconds
00:00:13.03 kernel: Performing SHA256 hash...
00:00:28.01 kernel: SHA256 sum for /testfile.bin: 7f8b25dd6fd03f90d7e903b7b07b07d9a758c41cefb7e656a258eec3d4836c78
00:00:28.02 kernel: SHA256 sum MATCH
00:00:28.03 kernel: Writing 265MB to USB:...
00:00:28.14 dwhci: Transaction failed (status 0x202)
00:00:28.15 dwhci: Transaction failed (status 0x202)
00:00:28.41 dwhci: Transaction failed (status 0x202)
00:00:28.42 dwhci: Transaction failed (status 0x202)
00:00:28.68 dwhci: Transaction failed (status 0x202)
00:00:29.19 dwhci: Transaction failed (status 0x202)
00:00:29.45 dwhci: Transaction failed (status 0x202)
00:00:29.71 dwhci: Transaction failed (status 0x202)
00:00:29.97 dwhci: Transaction failed (status 0x202)
00:00:30.24 dwhci: Transaction failed (status 0x202)
00:00:30.50 dwhci: Transaction failed (status 0x202)
00:00:30.76 dwhci: Transaction failed (status 0x202)
00:00:31.27 dwhci: Transaction failed (status 0x202)
00:00:31.53 dwhci: Transaction failed (status 0x202)
00:00:31.80 dwhci: Transaction failed (status 0x202)
00:00:32.06 dwhci: Transaction failed (status 0x202)
00:00:32.57 dwhci: Transaction failed (status 0x202)
00:00:32.83 dwhci: Transaction failed (status 0x202)
00:00:33.09 dwhci: Transaction failed (status 0x202)
00:00:33.36 dwhci: Transaction failed (status 0x202)
00:00:33.62 dwhci: Transaction failed (status 0x202)
00:00:33.88 dwhci: Transaction failed (status 0x202)
00:00:34.14 dwhci: Transaction failed (status 0x202)
00:00:34.40 dwhci: Transaction failed (status 0x202)
00:00:34.91 dwhci: Transaction failed (status 0x202)
00:00:35.18 dwhci: Transaction failed (status 0x202)
00:00:35.44 dwhci: Transaction failed (status 0x202)
00:00:36.21 dwhci: Transaction failed (status 0x202)
00:00:36.47 dwhci: Transaction failed (status 0x202)
00:00:36.73 dwhci: Transaction failed (status 0x202)
00:00:36.99 dwhci: Transaction failed (status 0x202)
00:00:37.25 dwhci: Transaction failed (status 0x202)
00:00:37.26 dwhci: Transaction failed (status 0x202)
00:00:37.26 dwhci: Transaction failed (status 0x202)
00:00:37.27 dwhci: Transaction failed (status 0x202)
00:00:37.27 dwhci: Transaction failed (status 0x202)
00:00:37.28 dwhci: Transaction failed (status 0x202)
00:00:37.28 dwhci: Transaction failed (status 0x202)
00:00:37.29 dwhci: Transaction failed (status 0x202)
00:00:37.29 dwhci: Transaction failed (status 0x202)
00:00:37.29 dwhci: Transaction failed (status 0x202)
00:00:37.30 dwhci: Transaction failed (status 0x202)
00:00:37.82 dwhci: Transaction failed (status 0x202)
00:00:38.08 dwhci: Transaction failed (status 0x202)
00:00:38.59 dwhci: Transaction failed (status 0x202)
00:00:38.85 dwhci: Transaction failed (status 0x202)
00:00:39.12 dwhci: Transaction failed (status 0x202)
00:00:39.12 dwhci: Transaction failed (status 0x202)
00:00:39.12 dwhci: Transaction failed (status 0x202)
00:00:39.13 dwhci: Transaction failed (status 0x202)
00:00:39.13 dwhci: Transaction failed (status 0x202)
00:00:39.14 dwhci: Transaction failed (status 0x202)
00:00:39.14 dwhci: Transaction failed (status 0x202)
00:00:39.15 dwhci: Transaction failed (status 0x202)
00:00:39.41 dwhci: Transaction failed (status 0x202)
00:00:39.67 dwhci: Transaction failed (status 0x202)
00:00:39.93 dwhci: Transaction failed (status 0x202)
00:00:40.19 dwhci: Transaction failed (status 0x202)
00:00:40.45 dwhci: Transaction failed (status 0x202)
00:00:40.71 dwhci: Transaction failed (status 0x202)
00:00:40.97 dwhci: Transaction failed (status 0x202)
00:00:41.23 dwhci: Transaction failed (status 0x202)
00:00:41.75 dwhci: Transaction failed (status 0x202)
00:00:42.01 dwhci: Transaction failed (status 0x202)
00:00:42.27 dwhci: Transaction failed (status 0x202)
00:00:42.79 dwhci: Transaction failed (status 0x202)
00:00:43.05 dwhci: Transaction failed (status 0x202)
00:00:43.57 dwhci: Transaction failed (status 0x202)
00:00:43.83 dwhci: Transaction failed (status 0x202)
00:00:44.09 dwhci: Transaction failed (status 0x202)
00:00:44.35 dwhci: Transaction failed (status 0x202)
00:00:44.61 dwhci: Transaction failed (status 0x202)
00:00:44.87 dwhci: Transaction failed (status 0x202)
00:00:45.13 dwhci: Transaction failed (status 0x202)
00:00:45.39 dwhci: Transaction failed (status 0x202)
00:00:45.65 dwhci: Transaction failed (status 0x202)
00:00:45.91 dwhci: Transaction failed (status 0x202)
00:00:46.17 dwhci: Transaction failed (status 0x202)
00:00:46.44 dwhci: Transaction failed (status 0x202)
00:00:46.61 kernel: File written in 18.58 seconds
00:00:46.62 kernel: === TEST PASSED! (1/1) ===
00:00:46.62 kernel: Remove USB flash drive!
00:00:52.84 usbdev0-1: Device ven781-5581 removed
00:00:52.85 kernel: Plug in an USB flash drive!
00:00:55.05 usbdev0-1: Device ven781-5581 found
00:00:55.05 usbdev0-1: Interface int8-6-50 found
00:00:55.06 usbdev0-1: Using device/interface int8-6-50
00:00:55.22 umsd: Capacity is 58656 MByte
00:00:55.23 usbhub: Port 2: Device configured
00:00:55.32 kernel: Reading 265MB from USB:...
00:01:04.01 kernel: File read in 8.68 seconds
00:01:04.02 kernel: Performing SHA256 hash...
00:01:19.01 kernel: SHA256 sum for /testfile.bin: 7f8b25dd6fd03f90d7e903b7b07b07d9a758c41cefb7e656a258eec3d4836c78
00:01:19.02 kernel: SHA256 sum MATCH
00:01:20.08 kernel: Writing 265MB to USB:...
00:01:37.64 kernel: File written in 17.55 seconds
00:01:37.64 kernel: === TEST PASSED! (2/2) ===
00:01:37.65 kernel: Remove USB flash drive!
00:01:42.77 usbdev0-1: Device ven781-5581 removed
00:01:42.78 kernel: Plug in an USB flash drive!
00:01:44.47 usbdev0-1: Device ven781-5581 found
00:01:44.48 usbdev0-1: Interface int8-6-50 found
00:01:44.48 usbdev0-1: Using device/interface int8-6-50
00:01:44.64 umsd: Capacity is 58656 MByte
00:01:44.65 usbhub: Port 2: Device configured
00:01:44.67 kernel: Reading 265MB from USB:...
00:01:53.36 kernel: File read in 8.68 seconds
00:01:53.36 kernel: Performing SHA256 hash...
00:02:08.36 kernel: SHA256 sum for /testfile.bin: 7f8b25dd6fd03f90d7e903b7b07b07d9a758c41cefb7e656a258eec3d4836c78
00:02:08.37 kernel: SHA256 sum MATCH
00:02:09.22 kernel: Writing 265MB to USB:...
00:02:26.89 kernel: File written in 17.66 seconds
00:02:26.90 kernel: === TEST PASSED! (3/3) ===
00:02:26.90 kernel: Remove USB flash drive!
...
```

#### Non-PnP kernel (OK)
```
logger: Circle 43.3 started on Raspberry Pi 3 Model B (AArch64)
00:00:00.66 timer: SpeedFactor is 1.51
00:00:00.69 sdhost: emmc1: sdhost-bcm2835 loaded
00:00:02.20 emmc: Found a valid version 3.0x SD card
00:00:03.06 usbdev0-1: Device ven424-9514, dev9-0-2 found
00:00:03.07 usbdev0-1: Interface int9-0-1 found
00:00:03.07 usbdev0-1: Function is not supported
00:00:03.08 usbdev0-1: Interface int9-0-2 found
00:00:03.08 usbdev0-1: Using device/interface int9-0-2
00:00:03.80 usbdev0-1: Device ven424-ec00 found
00:00:03.80 usbdev0-1: Using device/interface ven424-ec00
00:00:03.96 usbdev0-1: Device ven781-5581 found
00:00:03.97 usbdev0-1: Interface int8-6-50 found
00:00:03.97 usbdev0-1: Using device/interface int8-6-50
00:00:04.03 smsc951x: MAC address is B8:27:EB:78:10:CD
00:00:04.04 usbhub: Port 1: Device configured
00:00:04.19 umsd: Capacity is 58656 MByte
00:00:04.20 usbhub: Port 2: Device configured
00:00:04.21 dwroot: Device configured
00:00:04.21 kernel: Compile time: Mar 10 2021 22:51:10
00:00:04.22 kernel: Reading 265MB from USB:...
00:00:12.91 kernel: File read in 8.68 seconds
00:00:12.91 kernel: Performing SHA256 hash...
00:00:27.87 kernel: SHA256 sum for /testfile.bin: 7f8b25dd6fd03f90d7e903b7b07b07d9a758c41cefb7e656a258eec3d4836c78
00:00:27.88 kernel: SHA256 sum MATCH
00:00:27.89 kernel: Writing 265MB to USB:...
00:00:45.61 kernel: File written in 17.71 seconds
00:00:45.61 kernel: === TEST PASSED! (1/1) ===
00:00:45.63 kernel: Reading 265MB from USB:...
00:00:54.32 kernel: File read in 8.68 seconds
00:00:54.32 kernel: Performing SHA256 hash...
00:01:09.29 kernel: SHA256 sum for /testfile.bin: 7f8b25dd6fd03f90d7e903b7b07b07d9a758c41cefb7e656a258eec3d4836c78
00:01:09.30 kernel: SHA256 sum MATCH
00:01:10.70 kernel: Writing 265MB to USB:...
00:01:29.07 kernel: File written in 18.36 seconds
00:01:29.07 kernel: === TEST PASSED! (2/2) ===
00:01:29.08 kernel: Reading 265MB from USB:...
00:01:37.78 kernel: File read in 8.68 seconds
00:01:37.78 kernel: Performing SHA256 hash...
00:01:52.74 kernel: SHA256 sum for /testfile.bin: 7f8b25dd6fd03f90d7e903b7b07b07d9a758c41cefb7e656a258eec3d4836c78
00:01:52.75 kernel: SHA256 sum MATCH
00:01:54.36 kernel: Writing 265MB to USB:...
00:02:53.05 kernel: File written in 58.67 seconds
00:02:53.05 kernel: === TEST PASSED! (3/3) ===
...
```

So, to summarize:
- 3A+ has trouble initializing the stick with this test kernel in PnP and Non-PnP.
  * In my actual application, which has much more going on (multi core, timer interrupts, audio), the stick *IS* initialized and can be read for some time, but will usually fail with a similar error halfway through reading a second file.
- 3B complains of Transaction Failed with PnP kernel, even though the read data integrity is good. Seem OK in Non-PnP kernel.

I'm not sure how the test performs on a 3 B**+**, as unfortunately my 3 B+ was accidentaly damaged a few days ago, but I was wondering if you are able to reproduce any of the above to start with.

Thanks again for your time!

**Comment by rsta2** [comment]
Thanks for the report and the valuable test kernel!

First I tried it on a RPi 3B+ and it behaves like the RPi 3B (a number of "Transaction failed (status 0x202)" messages, but transfers are OK). These messages come from polling the hub controller of the on-board hub(s) of the RPi 3B(+) for plug-and-play events and are not serious. They can occur and the hub driver is restarting the transfer silently, if an transaction fails with this status.

Now, because I do not have an RPi 3A+, I tested your test kernel on an RPi Zero W and it shows the same behavior like the RPi 3A+. What makes the difference here is the USB hub, which is integrated on the RPi 3B(+), but not on the RPi 3A+ and RPi Zero W. All these RPi models have the same USB host controller. I attached an external USB hub to the RPi Zero W and with the USB flash drive connected to this external hub, all transfers are OK.

Of course the aborted transfers are serious and I think, I can analyze this further using the RPi Zero W, which will take some time.

**Comment by dwhinham** [comment]
Thankyou Rene, as always, take as much time as you need 🙂

I guessed that the differences would be because of the lack of the integrated hub on the Pi 3A+, so a good idea to to test on the Zero W, I didn't think of that!

Regardless, a Pi 3 A+ is on its way to you from a supplier in Germany and it should hopefully be with you in a few days - even if you manage to find a solution without it, I hope it might be useful to you; it's a nice small form factor.

Cheers! 🙂

**Comment by rsta2** [comment]
Thank you very much, Dale!

**Comment by rsta2** [comment]
Hi Dale, I corrected the hard errors in reset recovery and now your test kernel works better on the RPi 3A+. One may still get error messages, but in most cases the driver can recover from these conditions. Unfortunately this still does not work in all cases yet. It is probably necessary to totally suppress these error conditions in the DWHCI driver and I tried several things, but wasn't successful so far. Unfortunately there will probably not a quick fix for this issue.

**Comment by rsta2** [comment]
Yes, I guess, it's a timing issue. But I was playing with some parameters in this direction without success. It's also lack of knowledge. The Linux driver, which I have analyzed to implement the Circle driver, was rather complex. I may have missed something.

**Comment by rsta2** [comment]
Now I found a way to retry failed USB bulk transactions on a low level, so that the storage driver will not notice it. Your test kernel did run here one hour without any problem with a SanDisk Ultra Flair USB 3.0 32GB flash drive with this modification on a RPi 3A+. I hope, this fixes this issue for you too. Can you please test it. Thanks! The patch is on the *develop* branch.

**Comment by dwhinham** [comment]
Hi Rene,

This is amazing - I think you've done it. I didn't even bother with the test kernel and I applied these new commits directly to my project. Thankyou so much!

USB mass storage now seems to be stable with my 3A+ and my 3B - I can load large files from USB stick reliably.
As you mentioned, I still get "Transaction failed" on the 3B, but the file loads successfully as expected. On the 3A+ the log is always clean!

I've decided to release a new version of my project with USB storage as an "experimental" feaature. That way we can get more feedback as I have quite a number of active users.

Already though this is looking very, very promising. Thankyou for the time you have spent looking into this!
Have a great weekend! 😃

**Comment by rsta2** [comment]
Hi Dale, thank you, also for testing! I appreciate that you have created a release including these modifications. The feedback from your users will be very valuable. Have a great weekend too!

**Comment by philtimmes** [comment]
Could you suppress (disable) plug&play polling during active transactions, and then re-enable them afterwards?
_EnterFileOpenCritical();
_ExitFileOpenCritical();
??

**Comment by rsta2** [comment]
@philtimmes For what purpose should I do this?

**Comment by rsta2** [comment]
This issue has been fixed in Circle 44. So this can be closed.
